pkg/proto/dispatch/v1/01_codec.go

```diff
@@ -20,38 +18,51 @@ import (
 const Name = "proto"
 
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToSizedBufferVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
 }
 
-type vtprotoCodec struct{}
+type vtprotoCodec struct {
+	fallback encoding.CodecV2
+}
 
 func (vtprotoCodec) Name() string { return Name }
 
-func (vtprotoCodec) Marshal(v any) ([]byte, error) {
+func (c *vtprotoCodec) Marshal(v any) (mem.BufferSlice, error) {
 	if m, ok := v.(vtprotoMessage); ok {
-		return m.MarshalVT()
-	}
-
-	if m, ok := v.(proto.Message); ok {
-		return proto.Marshal(m)
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
+			if _, err := m.MarshalToSizedBufferVT(buf[:size]); err != nil {
+				return nil, err
+			}
+			return mem.BufferSlice{mem.SliceBuffer(buf)}, nil
+		}
+		pool := mem.DefaultBufferPool()
+		buf := pool.Get(size)
+		if _, err := m.MarshalToSizedBufferVT((*buf)[:size]); err != nil {
+			pool.Put(buf)
+			return nil, err
+		}
+		return mem.BufferSlice{mem.NewBuffer(buf, pool)}, nil
```

Comment:
Curious - what does `NewBuffer` do here? Haven't we already reserved the relevant memory in the `pool.Get` call above?

Alternatively, is there some reading about how this approach works generally?

Reply:
By and large, this code is inspired by the upstream default codec: https://github.com/grpc/grpc-go/blob/v1.66.2/encoding/proto/proto.go

`mem.NewBuffer` is less about allocating bytes and more about tracking the lifetime and ownership of the buffer: https://github.com/grpc/grpc-go/blob/v1.66.2/mem/buffers.go#L83-L102